tests/cases/unittests/session.ts

```diff
@@ -0,0 +1,454 @@
+                });
+            });
+        });
+
+        describe("output", () => {
+            it("can format command responses and send them", () => {
+                const body = {
+                    block: {
+                        key: "value"
+                    }
+                };
+                const command = "test";
+
+                session.output(body, command);
+
+                expect(lastSent).to.deep.equal({
+                    seq: 0,
+                    request_seq: 0,
+                    type: "response",
+                    command,
+                    body: body,
+                    success: true
+                });
+            });
+        });
+    });
+
+    describe("how Session is extendable via subclassing", () => {
+        class TestSession extends Session {
+            lastSent: protocol.Message;
+            customHandler: string = "testhandler";
+            constructor(){
+                super(mockHost, Buffer.byteLength, process.hrtime, mockLogger);
+                this.addProtocolHandler(this.customHandler, () => {
+                    return {response: undefined, responseRequired: true};
+                });
+            }
+            send(msg: protocol.Message) {
+                this.lastSent = msg;
+            }
+        };
+
+        it("can override methods such as send", () => {
+            const session = new TestSession();
+            const body = {
+                block: {
+                    key: "value"
+                }
+            };
+            const command = "test";
+
+            session.output(body, command);
+
+            expect(session.lastSent).to.deep.equal({
+                seq: 0,
+                request_seq: 0,
+                type: "response",
+                command,
+                body: body,
+                success: true
+            });
+        });
+        it("can add and respond to new protocol handlers", () => {
+            const session = new TestSession();
+
+            expect(session.executeCommand({
+                seq: 0,
+                type: "command",
+                command: session.customHandler
+            })).to.deep.equal({
+                response: undefined,
+                responseRequired: true
+            });
+        });
+        it("has access to the project service", () => {
+            class ServiceSession extends TestSession {
+                constructor() {
+                    super();
+                    assert(this.projectService);
+                    expect(this.projectService).to.be.instanceOf(ProjectService);
+                }
+            };
+            new ServiceSession();
+        });
+    });
+
+    describe("an example of using the Session API to create an in-process server", () => {
+        class InProcSession extends Session {
+            private queue: protocol.Request[] = [];
+            constructor(private client: {handle: (msg: protocol.Message) => void}) {

```

Comment:
Oh, yeah. Sorry, I missed that one since it wasn't one of the intersection types.